src/mscorlib/src/System/String.CoreCLR.cs

```diff
@@ -0,0 +1,153 @@
+        }
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern static String FastAllocateString(int length);
+
+        // Is this a string that can be compared quickly (that is it has only characters > 0x80 
+        // and not a - or '
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern bool IsFastSort();
+        // Is this a string that only contains characters < 0x80.
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern bool IsAscii();
+
+#if FEATURE_COMINTEROP
+        // Set extra byte for odd-sized strings that came from interop as BSTR.
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern void SetTrailByte(byte data);
+        // Try to retrieve the extra byte - returns false if not present.
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern bool TryGetTrailByte(out byte data);
+#endif
+
+        public static String Intern(String str)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+
+            return Thread.GetDomain().GetOrInternString(str);
+        }
+
+        public static String IsInterned(String str)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+
+            return Thread.GetDomain().IsStringInterned(str);
+        }
+
+        // Copies the source String (byte buffer) to the destination IntPtr memory allocated with len bytes.
+        internal unsafe static void InternalCopy(String src, IntPtr dest, int len)
+        {
+            if (len == 0)
+                return;
+            fixed (char* charPtr = &src._firstChar)
+            {
+                byte* srcPtr = (byte*)charPtr;
+                byte* dstPtr = (byte*)dest;
+                Buffer.Memcpy(dstPtr, srcPtr, len);
+            }
+        }
+
+        unsafe internal int GetBytesFromEncoding(byte* pbNativeBuffer, int cbNativeBuffer, Encoding encoding)
+        {
+            // encoding == Encoding.UTF8
+            fixed (char* pwzChar = &_firstChar)
+            {
+                return encoding.GetBytes(pwzChar, Length, pbNativeBuffer, cbNativeBuffer);
+            }
+        }
+
+        unsafe internal int ConvertToAnsi(byte* pbNativeBuffer, int cbNativeBuffer, bool fBestFit, bool fThrowOnUnmappableChar)
+        {
+            Debug.Assert(cbNativeBuffer >= (Length + 1) * Marshal.SystemMaxDBCSCharSize, "Insufficient buffer length passed to ConvertToAnsi");
+
+            const uint CP_ACP = 0;
+            int nb;
+
+            const uint WC_NO_BEST_FIT_CHARS = 0x00000400;
+
+            uint flgs = (fBestFit ? 0 : WC_NO_BEST_FIT_CHARS);
+            uint DefaultCharUsed = 0;
+
+            fixed (char* pwzChar = &_firstChar)
+            {
+                nb = Win32Native.WideCharToMultiByte(
```

Comment:
Elsewhere when MultiByteToWideChar was used, Encoding.UTF8.GetBytes was used instead on Unix.  Is there something to be done here that's the equivalent?

Reply:
Yes. Ideally, methods like this one would be moved to where interop-specific helpers live. They do not belong to String.cs. Similar to how it is done in CoreRT, and shared with CoreRT.

There are subtle details to figure out and test, e.g. how to map `flags` to the right flavor of UTF8 encoding. For now, I just copied these verbatim from String.cs to String.CoreCLR.cs without changing anything. I do not think we have good enough test coverage for these to be able to touch them with confidence.